elki-core-math/src/main/java/elki/math/statistics/tests/mcde/MWPTest.java

```diff
@@ -0,0 +1,191 @@
+   *
+   * @param adapter ELKI NumberArrayAdapter Subclass
+   * @param data    One dimensional array containing one dimension of the data
+   * @param idx     Return value of sortedIndex()
+   * @return Array of doubles, 3 subsequent values being assigned to one data instance.
+   * Containing sorted (ascending) row numbers, adjusted ranks and tying value corrections
+   * as required by MWP test. Example:
+   * double[] corrected_ranks = corrected_ranks(...);
+   * double l = corrected_rank[0]; double adjusted_rank = corrected_rank[1]; double correction = corrected_rank[2];
+   * <p>
+   * Correspond to one instance of the original data
+   */
+  public <A> MwpIndex[] corrected_ranks(final NumberArrayAdapter<?, A> adapter, final A data, int[] idx) {
+    final int len = adapter.size(data);
+    MwpIndex[] I = (MwpIndex[]) Array.newInstance(MwpIndex.class, len);
+
+    int j = 0;
+    long correction = 0;
+    while(j < len) {
+      int k = j;
+      int t = 1;
+      double adjust = 0.0;
+
+      while((k < len - 1) && (adapter.getDouble(data, idx[k]) == adapter.getDouble(data, idx[k + 1]))) {
+        adjust += k;
+        k++;
+        t++;
+      }
+
+      if(k > j) {
+        double adjusted = (adjust + k) / t;
+        correction += (t * t * t) - t;
+
+        for(int m = j; m <= k; m++) {
+          I[m] = new MwpIndex(idx[m], adjusted, correction);
+        }
+      }
+      else
+        I[j] = new MwpIndex(idx[j], j, correction);
+      j += t;
+    }
+
+    return I;
+  }
+
+  /**
+   * Efficient implementation of MWP statistical test using appropriate index structure as described in Algorithm 3
+   * of reference paper. Tailored to MCDE Framework.
+   *
+   * @param start           Starting index value for statistical test
+   * @param end             End index value for statistical test
+   * @param slice           Return value of randomSlice() created with the index that is not for the reference dimension
+   * @param corrected_ranks Index of the reference dimension, return value of corrected_ranks() computed for reference dimension
+   * @return p-value from two sided Mann-Whitney-U test
+   */
+  public double statistical_test(int start, int end, boolean[] slice, MwpIndex[] corrected_ranks) {
+    double R = 0.0;
+    long n1 = 0;
+    for(int j = start; j < end; j++) {
+
+      if(slice[corrected_ranks[j].index]) {
+        R += corrected_ranks[j].adjusted;
+        n1++;
+      }
+    }
+
+    // This is to cancel the offset in case the marginal restriction does not start from 0
+    // see "acc - (cutStart * count)" is reference implementation of MWP
+    R -= start * n1;
+
+    final int cutLength = end - start;
+    if((n1 == 0) || (n1 == cutLength)) {
+      return 1;
+    }
+
+    final double U = R - ((double) (n1 * (n1 - 1))) / 2;
+    final long n2 = cutLength - n1;
+
+    if(n1 + n2 > two_times_sqrt_max_long) {
```

Comment:
By the definition of `n2`, `n1 + n2` always is `cutLength`. Because `cutLength` is integer, it will always be less than this constant?

Reply:
This mechanism was supposed to secure the user against using too many instances resulting in `n1*n2` being too large to fit into a `long` and thus resulting in a negative number which in turn would result in returning `NaN` after the MCDE runs. The user then does not understand why `NaN` is returned because he assumes that everything he does is correct. I have taken over this from out Scala implementation where it was tested and worked well. However from looking at the code I see your point as `cutLength` as well as `start` and `end` are all ints. I will look at it more closely soon and get back to you about it if necessary.

Reply:
I just learned that arrays in Java cannot be larger than `Integer.MAX_VALUE`. In that case I would suggest to simply convert  `n1` and `n2` into a `double`, then there should be no issue with `n1*n2` exceeding the max limit of the data type. Note that the double cast when computing `std` will not be required anymore.